### ISSUE TITLE
CTC-segmentation of large corpora

### ISSUE THREAD

**Issue by lumaku** [espnet/espnet#2192]
In our work, we used the segmentation with CTC as originally proposed by Graves et al. in his original publication on CTC (for protein structures etc.).

With that, our approach to CTC-segmentation uses an ASR model pre-trained with CTC to boot-strap more training data from unsegmented or unlabeled training data.
Utterances are then extracted from label probabilities obtained from the network trained with CTC to determine segment alignments.
We used this to generate over 1700h of labeled speech training data, see https://arxiv.org/abs/2007.09127 .

The source code for this method can be found here: https://github.com/cornerfarmer/ctc_segmentation
Please advise on how to merge this with ESPnet, e.g., by adding it to `espnet/utils`.

**Comment by sw005320** [comment]
This is very impressive work (and quite timely for me, @jnishi, and @simpleoier).
Many thanks for introducing it!
And, yes, we definitely want to include it in espnet.
(As you may notice), we recently made a prototype [alignment tool](https://github.com/espnet/espnet/blob/master/utils/ctc_align_wav.sh), but your tool is more complete.

What I'm thinking is to include it in `espnet/utils` as you mentioned or regarding it as a recipe. 
The current tool is rather complicated as a stand-alone tool, and my feeling is that we can put some of your scripts in `espnet/utils/align_xxx.{py.sh}` and then it is probably better to create some recipes, e.g., `egs/tedlium2/align1/`.
I and @jnishi have a similar project in Japanese, and we can follow or help your development.

@jnishi, could you also take a look at this tool and give us some comments?

**Comment by jnishi** [comment]
That sounds great. I agree with @sw005320.  It is good to add their core method to `espnet/utils` and align with pre-trained model listed in https://github.com/espnet/espnet#asr-demo. After that, it is good to add data preparation code as recipes to align with the model trained by users. 

@sw005320 I think the recipe of an alignment tool requires the corresponding asr recipe. Which is better that the alignment recipe have the (symbolic) link to the asr recipe or the alignment recipe includes asr training flow?

**Comment by lumaku** [comment]
@sw005320 Yes, I noticed your current work on `ctc_align`; that was also the motivation to post our work.

The CTC-Segmentation repository contains scripts to evaluate alignments and to compare them with MAUS and Gentle; these scripts would not be needed for usage in ESPnet. Starting from your current implementation of `ctc_align()` in `espnet/asr/pytorch_backend/asr.py`, roughly speaking, additional stuff from only two files in the CTC-segmentation repo is needed: the file containing the optimized alignment algorithm `align_fill.pyx` and `align.py` that contains the backtracking and labeling+segmentation.

The contribution could be done by either provide CTC-segmentation as separate tool similar to `ctc_align_wav.sh`;
or alternatively, to merge your current implementation of `ctc_align()` in `espnet/asr/pytorch_backend/asr.py` and extend it with ours in `ctc_segmentation/align.py`.

One remark about recipe organization (how we did it): We bootstrapped from the `tuda-de` recipe (ported from kaldi to ESPnet), and used this model to generate the large German corpus in a second step. The newly generated `german` corpus is then a different dataset that does not necessarily have an overlap with the former bootstrapping training dataset.

**Comment by sw005320** [comment]
Thanks for the detailed suggestions.
Our `ctc_align()` is an experimental function. So, we don’t have to maintain separate tools.
If merge is easy, it is better for me.
But it is no problem if we can just use yours and we can disable `ctc_align()`.

**Comment by sw005320** [comment]
> @sw005320 I think the recipe of an alignment tool requires the corresponding asr recipe. Which is better that the alignment recipe have the (symbolic) link to the asr recipe or the alignment recipe includes asr training flow?

This is a good point.
My preference is the first one (having the ASR recipe link).

**Comment by lumaku** [comment]
Then the plan is to port the CTC segmentation tool to ESPnet, while re-using existing components of `ctc_align()` and of `forced_align()` in `pytorch_backend/ctc.py`.

Regarding the timeline, I think that we can provide a pull request for this in about two or max. three weeks.

**Comment by sw005320** [comment]
Sounds very good. Please let us know if you want to have some questions and help about it.

**Comment by rajeevbaalwan** [comment]
@lumaku 
Can Force CTC alignment works for a batch instead alignment of 1 audio,text pair at a time ?

**Comment by lumaku** [comment]
Yes, this is possible.
Currently, a simple for loop in `espnet/bin/asr_align.py` iterates over audio files.

I'd propose this optimization:
The loop in `espnet/bin/asr_align.py` could be optimized in this way: In a for-loop, extract the `lpz` using the GPU and directly continue with the inference on the GPU of the next file; for each inferred file, dispatch an asyncio job on a different thread to apply `ctc_segmentation()` and process their results.

There are two computationally-heavy bottlenecks:
1. In a first step, the CTC activations are calculated on `device` (GPU for ~10x-60x speedup), i.e., `lpz` - This could be done in batch. Please consider that audio files processed in a single batch might need to have around the same length (depends on the implementation of the padding and the network, ....).

2. The second step calculates the foward-backward tracking on the CPU, i.e. `ctc_segmentation()`. The computation only needs the CTC activations of the previous step to start and one file is processed at a time. This is currently done in a single thread which can be parallelized - mainly limited by how much RAM you have on your computer (to store intermediate data, CTC activations etc).

Currently, only decoding with one GPU is implemented; if you have multiple GPUs, you might as well partition your dataset and start different instances of `asr_align.py`.

**Comment by rajeevbaalwan** [comment]
Thanks @lumaku for reponse 
I have already implemented required changes upto 1st step. I was trying to implement 2nd step on a batch in 1 go instead of running parallel threads. That's where I get stuck. 😶
Will try the parallel thread approach.

**Comment by lumaku** [comment]
The current `ctc_segmentation()` function aligns utterances of a single file - it only requires a single CPU thread.
If you already inferred the files in a batch, a possible next step could be to partition the batch and distribute it to concurrent `ctc_segmentation()` jobs.

**Comment by rajeevbaalwan** [comment]
> The current `ctc_segmentation()` function aligns utterances of a single file - it only requires a single CPU thread.
> If you already inferred the files in a batch, a possible next step could be to partition the batch and distribute it to concurrent `ctc_segmentation()` jobs.

Yes that's certainly a way to go, but I'm trying to explore the possibility to tweak ctc_segmentation implementation or some part of it to make it work on utterances of batch  instead of utterances of single file which will be fast.

**Comment by lumaku** [comment]
Regarding the current code path and optimizations, the current implementation is more suitable for single thread CPU calculations. For true batch processing, an adaption to conduct the CTC segmentation on GPU would certainly be feasible. There are batched GPU implementations of the forward-backward algorithm, and I assume that a batched CTC segmentation on the GPU would be similar.

Memory consumption is one thing to keep in mind, as the CTC activations and the table of probabilities (frames x text) are kept in memory. In our work, we segmented audio files of up to several hours.

**Comment by rajeevbaalwan** [comment]
> Regarding the current code path and optimizations, the current implementation is more suitable for single thread CPU calculations. For true batch processing, an adaption to conduct the CTC segmentation on GPU would certainly be feasible. There are batched GPU implementations of the forward-backward algorithm, and I assume that a batched CTC segmentation on the GPU would be similar.
> 
> Memory consumption is one thing to keep in mind, as the CTC activations and the table of probabilities (frames x text) are kept in memory. In our work, we segmented audio files of up to several hours.

@lumaku 
I haven't found batched GPU implementation of forward backward algorithm. Can you please where I can got the  Batched GPU implementation of backward forward algorithm or ctc segmentation.

only True Batched CTC segmentation can solve my problem as I have tried parallel processing and these things I noticed:

Firstly multi threading doesn't improve the performance when it comes to large batch size processing like 32, 64 as the task per thread is CPU bound infect it degrades the performance and can't match up the models throughput.

Secondly as task are CPU bound multiprocessing improved the processing time for large batch size but at cost of CPU load average which also will not work for me as I have 8 GPU machine. When I process batch of size 32 or 64 my CPU load jumps to range 9-12 for only single GPU. So I can't run multiprocessing also as it would increase CPU load by  8X  which obviously I can't provide.

That's why I needed to implement batch CTC segmentation

**Comment by lumaku** [comment]
I see. For this use case, CTC segmentation would have to be re-implemented for the GPU. There are papers that discuss possible optimizations for similar algorithms, e.g., [this Google paper](https://ieeexplore.ieee.org/abstract/document/8268944).

CTC segmentation is usually only done once. Consider distributing the computation load on multiple computers using a workload manager.

If you decide to implement this on the GPU, please let me know; also, I can integrate the GPU implementation into the python package.

**Comment by sw005320** [comment]
Can we refer or reuse the batch implementation of CTC prefix score https://github.com/espnet/espnet/blob/master/espnet/nets/ctc_prefix_score.py#L12 ?
This includes the CTC forward computation.

**Comment by lumaku** [comment]
The CTC prefix score implements a slightly modified forward algorithm. From the [algorithm description (Alg. 2)](https://ieeexplore.ieee.org/abstract/document/8068205/) and the implementation, it seems that the table of probabilities is discarded, and only the batch-forward probability `log_psi` is returned.

CTC prefix scoring shares similarities with `cython_fill_table()` in [`ctc_segmentation_dyn.pyx`](https://github.com/lumaku/ctc-segmentation/blob/master/ctc_segmentation/ctc_segmentation_dyn.pyx). There are some minor differences, e.g., `cython_fill_table` keeps the transition cost zero before the start of the utterance.

However, there is one further requirement: Intermediate forward probabilities are stored in a table in order to do the backward pass. This shares similarity to the CTC loss algorithm. Are there any batched pytorch implementations of CTC to use as a blueprint for an implementation?

@rajeevbaalwan For a GPU implementation, `cython_fill_table()` would be a good function to start the optimization, as most of the computation of `ctc_segmentation()` happens within this function. Also, the current implementation uses Numpy. Those computations are done in sequence (each forward probability iteration depends on the previous time step!). For a naive port to Pytorch, a GPU implementation would need to vectorize the calculation of forward probabilities (similar as in `CTCPrefixScoreTH`) in order to benefit from parallelization.

**Comment by jnishi** [comment]
If python implementation of CTC is needed, how about that of Chainer.
It has vectorized computation of the forward-backward algorithm for GPU. It also supports the mini-batch computation of variable length examples.
It is a python implementation that uses Numpy and CuPy, which is the GPU implementation of Numpy.

https://github.com/chainer/chainer/blob/master/chainer/functions/loss/ctc.py

**Comment by lumaku** [comment]
Thank you for this hint. In Chainer, the [forward and backward pass](https://github.com/chainer/chainer/blob/13a016d1bdb12815ff737f58a74a20768003c305/chainer/functions/loss/ctc.py#L212) are both implemented in Cuda code. This function could very well be adapted for CTC segmentation.

However, I also think that the current implementation on the CPU is sufficient. The larger the segmented audio files, the less an implementation could benefit from the advantages of efficient parallelization on the GPU. CTC segmentation is a sequential algorithm and therefore more suitable for the CPU. Memory on the GPU is limited and therefore would only allow limited batch sizes (although subsampling in the encoder helps to reduce memory consumption). Furthermore, single core tasks on the GPU perform slower than on the CPU.
